### ISSUE TITLE
plz is extremely busy... should it be?

### ISSUE THREAD

**Issue by phil-s** [alphapapa/ement.el#197]
If I `(trace-function 'plz--respond)` I'm seeing about 5 calls per second consistently.

That really seems like a lot.  (Especially if each of those is starting an independent curl process?)

Is that level of activity in line with expectations?

**Comment by phil-s** [comment]
I guess `ement-auto-sync` just fires off each request as soon as it's handled the previous one?

I reckon that needs some kind of throttling.

In the meantime I've disabled that and am instead running https://github.com/alphapapa/ement.el/issues/177#issuecomment-1609426587 on a shorter interval than the one in that code.

**Comment by alphapapa** [comment]
Matrix is implemented via an HTTP connection that remains open until the server has data to send to the client.  That data might be a large chunk of events for multiple rooms, or it might be a tiny response indicating that a user is typing in a room.  As soon as such a response is received, the client makes another request to listen for the next update.  If the user is in several busy rooms, it's entirely possible, if not very likely, that multiple such requests and responses will be made in short periods of time.

Ement already limits to one outstanding sync per session: https://github.com/alphapapa/ement.el/blob/60e9a50b8a230c00d92f639ba25040944bcda9ab/ement.el#L492-L499

However, requests other than sync requests could be outstanding, e.g. sending messages, uploading files, retrieving images, etc.  All of those may run in parallel.

> I expect that some kind of customizable throttling would be a good idea? I'm imagining something like "poll every X seconds; but when a message is received (or sent!) then temporarily reduce the polling interval to Y seconds so that active conversations are processed in a timely fashion. Maintain the smaller value until such time as there have been no new messages for Z seconds, at which point revert the interval back to X." (With X, Y, Z being user options.)

I don't know of any reason such a thing would be needed.  AFAIK incoming events are processed correctly and quickly already.  This has worked fine ever since the first iteration of this client.

What is the problem you're reporting?

**Comment by phil-s** [comment]
> Matrix is implemented via an HTTP connection that remains open until the server has data to send to the client.

Ah, good, that make things less alarming!

> What is the problem you're reporting?

My real problem was Emacs continually maxing out a CPU core, and when I profiled that I got this:

```plain
       79180  65% - timer-event-handler
       79172  65%  - apply
       78361  64%   - plz--respond
       78330  64%    - #<compiled -0x187abbe156b59d04>
       77275  63%     - #<compiled -0xade032038805cf0>
       77275  63%      - apply
       77275  63%       - ement--sync-callback
       77229  63%        - run-hook-with-args
       76069  62%         - ement-room-list-auto-update
       76065  62%          - revert-buffer
       76065  62%           - ement-room-list-revert
       76045  62%            - ement-room-list
       28118  23%             + taxy-magit-section-format-items
       21583  17%             + taxy-magit-section-insert
       21551  17%             + taxy-fill
        3366   2%             + taxy-sort
         735   0%             + taxy-make-take-function
          80   0%             + walk-windows
```

Disabling `ement-room-list-auto-update` helped of course, but (I'm pretty sure) I didn't see the CPU usage completely drop away until I implemented my makeshift throttling.

I am joined to a *lot* of rooms in my primary Matrix server (~180), so that's a huge factor, but it would be good if the user had a way to help ement/emacs cope better with heavier loads like this.

**Comment by alphapapa** [comment]
I see, thanks.  Well, I'm in 103 rooms, and I don't have any performance problems with the room list updating, so I'm surprised that you're seeing that.  From looking at the profile, it doesn't look like `walk-windows` is a factor, but I still wonder, if you remove that patch that does that (i.e. run the latest release of Ement unmodified), does the performance problem remain?

Anyway, if necessary, the room list auto update function could be modified to run a timer of something like 200ms and reset it each time, so that if the function is called in rapid succession, the room list would only get updated when 200ms had expired since the last time it was called; that could serve to "debounce" the updates.

**Comment by phil-s** [comment]
> I'm in 103 rooms, and I don't have any performance problems with the room list updating, so I'm surprised that you're seeing that.

Ah, darn... I'd figured my case might be abnormal among ement.el users, but a factor of < 2x isn't so huge.  Or not unless there's some exponential effects which are starting to really ramp up around the upper half of that range.

I guess I'll need to try to learn more.  Which probably won't happen in a hurry, but leave it with me and I'll follow up when I have more info (and feel free to close this, I guess -- it can always be re-opened).

**Comment by phil-s** [comment]
I've been alerted to a Matrix server bug which is probably one (or both) of the following, which was apparently resulting in an enormous increase in Matrix traffic on our network.

* https://github.com/matrix-org/synapse/issues/15980
* https://github.com/matrix-org/synapse/issues/16057

I strongly suspect what I was seeing yesterday was simply on account of this, as the timing fits.

I'll re-test with defaults later on (as presence has now been disabled for our server to work around this issue).

**Comment by phil-s** [comment]
I haven't gone right back to defaults, but I've reinstated `ement-auto-sync` and reverted the other things I did for this, and tracing `plz--respond` now gets me around 1-2 entries in a 30 second span.

> the room list auto update function could be modified to run a timer of something like 200ms and reset it each time, so that if the function is called in rapid succession, the room list would only get updated when 200ms had expired since the last time it was called; that could serve to "debounce" the updates.

Even though this case was a server problem, I think something like that would be a good idea, if only to protect against any similar cases in the future.

**Comment by alphapapa** [comment]
Upgrading to priority B since I heard from another user who might be experiencing this issue.

**Comment by phil-s** [comment]
As part of the testing/debugging for issue #298 I implemented the following:

https://github.com/phil-s/ement.el/commit/6e696d56d2b32710d9e39d80019bdc99a57e4031

It provides a couple of knobs for users to customise the room list refresh.

```el
(defcustom ement-room-list-update-interval 5
  "Minimum seconds between `ement-room-list-auto-update' refreshes.
If nil, then refresh immediately upon each sync.

In either case this is further subject to `ement-room-list-update-idle-delay'."
  :type '(choice (const :tag "Refresh immediately upon sync" nil)
                 (number :tag "Minimum seconds between refreshes")))

(defcustom ement-room-list-update-idle-delay 1
  "Necessary idle time before any `ement-room-list-auto-update' refresh.
If nil, then refresh immediately."
  :type '(choice (const :tag "Refresh immediately" nil)
                 (number :tag "Wait until idle for this many seconds")))
```

If both options are `nil` we get the current behaviour.

If only `ement-room-list-update-idle-delay` is nil then we get similar behaviour to the earlier patch here (i.e. updates in idle time).

If only `ement-room-list-update-interval` is nil then we don't wait for idle time to update, but we have a minimum number of seconds between updates.

If neither option is nil, then we have the minimum interval *and* then, once that time has elapsed, we also wait for idle time to update.

Because #298 has resolved the known performance issues, it's not obvious that we need this (and I'd *definitely* go with a nil `ement-room-list-update-interval` as a default); but it also might still be a nice-to-have in case of future issues (especially server-side issues, such as the one I experienced in the past which absolutely bombarded clients with events).

**Comment by phil-s** [comment]
I managed to mess up the option descriptions.  Now fixed.

**Comment by Thaodan** [comment]
The current code doesn't use the builtin debug logging and logs by default. After disabling that it's fine.
